### PR TITLE
Fix file reads in the prefetched data

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
## Summary
There is an error when reading bytes from log segment file to a ByteBuf in prefetched data. This PR fixes that error.

## Test
Unit test

### CODE REVIEW

File: ambry-utils/src/main/java/com/github/ambry/utils/Utils.java [hunk 1]
```diff
@@ -612,6 +612,35 @@ public static int readFileToByteBuffer(FileChannel fileChannel, long offset, Byt
...
+    // If the writable bytes in ByteBuf is less then the given size, ByteBuf would expand its capacity.
+    // Put a check here to make sure that we don't expand the capacity.
+    if (buffer.writableBytes() < size) {
+      throw new IllegalArgumentException(
+          "ByteBuf doesn't have enough writable bytes [" + buffer.writableBytes() + "<" + size + "]");
+    }
+    int bytesRead = 0;
+    while (bytesRead < size) {
+      int readSize = buffer.writeBytes(fileChannel, offset + bytesRead, size - bytesRead);
+      if (readSize == -1 || readSize == 0) {
```
Comment (JingQianCloud) [inline ambry-utils/src/main/java/com/github/ambry/utils/Utils.java:636]: In which case will we hit the EOF?
I thought when we read, we checked the getMessageInfo().getSize() already.

Comment (justinlin-linkedin) [reply ambry-utils/src/main/java/com/github/ambry/utils/Utils.java:636]: This is a generic method that might deal with other usage. That's why we check EOF here.

Comment (JingQianCloud) [reply ambry-utils/src/main/java/com/github/ambry/utils/Utils.java:636]: Then I assume we don't hit issue. The code is just to be safe. Right?

**Review by JingQianCloud — APPROVED** [review]

**Review by Arun-LinkedIn — APPROVED** [review]
Lgtm! Just a q.. what is the bug with current code?

**Comment by justinlin-linkedin** [comment]
> Lgtm! Just a q.. what is the bug with current code?

In some partition, reading a whole blob out would fail because file channel won't return all the required bytes at one read. And if we fail three times, this would mistakenly mark this partition unavailable.

**Pull request merged**